### PR TITLE
yii-widgets.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | https://github.com/yiisoft/view/issues/34

### CODE REVIEW

**Comment by terabytesoftw** [comment]
Tests pass correction in view type hint propierty https://github.com/yiisoft/view/pull/73.

**Comment by samdark** [comment]
@terabytesoftw merged that one...

**Comment by terabytesoftw** [comment]
> @terabytesoftw merged that one...

Thks, tests all pass.

File: composer.json [hunk 1]
```diff
@@ -1,39 +1,56 @@
 {
-    "name": "yiisoft/_____",
+    "name": "yiisoft/yii-widgets",
     "type": "library",
-    "description": "_____",
+    "description": "Colections widgets for Yii",
```
Comment (rustamwin) [inline composer.json]:
```suggestion
    "description": "Collections widgets for Yii",
```

File: src/Block.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * ```
+ *
+ * in subviews show block:
+ *
+ * <?= $this->getBlock('index') ?>
+ *
+ * Second parameter defines if block content should be outputted which is desired when rendering its content but isn't
+ * desired when redefining it in subviews.
+ */
+class Block extends Widget
```
Comment (rustamwin) [inline src/Block.php]: Could be final?

File: src/Breadcrumbs.php [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+ * You can use a view parameter (e.g. `$this->params['breadcrumbs']`) to configure the links in different views. In the
+ * layout view, you assign this view parameter to the {@see links} property like the following:
+ *
+ * ```php
+ * // $this is the view object currently being used
+ * echo Breadcrumbs::widget()
+ *     ->links() => isset($this->params['breadcrumbs']) ? $this->params['breadcrumbs'] : [];
+ * ```
+ */
+class Breadcrumbs extends Widget
```
Comment (rustamwin) [inline src/Breadcrumbs.php]: Could be final?

File: src/Breadcrumbs.php [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+     *
+     * @param array  $link the link to be rendered. It must contain the "label" element. The "url" element is optional.
+     * @param string $template the template to be used to rendered the link. The token "{link}" will be replaced by the
+     *                         link.
+     *
+     * @throws \InvalidArgumentException if `$link` does not have "label" element.
+     *
+     * @return string the rendering result
+     */
+    protected function renderItem(array $link, string $template): string
```
Comment (rustamwin) [inline src/Breadcrumbs.php]:
```suggestion
    private function renderItem(array $link, string $template): string
```

File: src/ContentDecorator.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ *     ->params([]),
+ *     ->view($this),
+ * ]) ?>
+ *
+ * some content here
+ *
+ * <?php ContentDecorator::end() ?>
+ * ```
+ */
+class ContentDecorator extends Widget
```
Comment (rustamwin) [inline src/ContentDecorator.php]: Could be final?

File: src/Block.php (hunk 2 above)
Comment (terabytesoftw) [reply src/Block.php]: Because they are base widgets that could be extended.

File: src/Breadcrumbs.php (hunk 3 above)
Comment (terabytesoftw) [reply src/Breadcrumbs.php]: Because they are base widgets that could be extended.

File: src/ContentDecorator.php (hunk 5 above)
Comment (terabytesoftw) [reply src/ContentDecorator.php]: Because they are base widgets that could be extended.

File: composer.json (hunk 1 above)
Comment (terabytesoftw) [reply composer.json]: Done.

File: src/Breadcrumbs.php (hunk 4 above)
Comment (terabytesoftw) [reply src/Breadcrumbs.php]: Class is extended in yii-bootstrap4, like the method.

File: src/Block.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+namespace Yiisoft\Yii\Widgets;
+
+use Yiisoft\View\WebView;
+use Yiisoft\Widget\Widget;
+
+/**
+ * Block records all output between {@see begin()} and {@see end()} calls and stores it in
+ * {@see \Yiisoft\View\View::$blocks}.
+ *
+ * {@see \Yiisoft\View\View} component contains two methods {@see \Yiisoft\View\View::beginBlock()} and
```
Comment (samdark) [inline src/Block.php]: There are such methods...

File: src/Block.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+    public function __construct(WebView $webview)
+    {
+        $this->webView = $webview;
+    }
+
+    /**
+     * Starts recording a block.
+     */
+    public function init(): void
```
Comment (samdark) [inline src/Block.php]: How about renaming it to `start()` or `capture()`?

File: src/Breadcrumbs.php [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+     * @var bool the first hyperlink in the breadcrumbs (called home link). If this property is true, it will default
+     * to a link pointing to HomeUrl '\' with the label 'Home'. If this property is false, the home link will
+     * not be rendered.
+     */
+    private bool $homeLink = true;
+
+    /**
+     * @var array $homeUrlLink
+     */
+    private array $homeUrlLink;
```
Comment (samdark) [inline src/Breadcrumbs.php]: Should be nullable and initialized with null or be an array with some default value.

File: src/ContentDecorator.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    private WebView $webView;
+
+    public function __construct(WebView $webview)
+    {
+        $this->webView = $webview;
+    }
+
+
+    public function init(): void
```
Comment (samdark) [inline src/ContentDecorator.php]: We'd better come up with a more descriptive name...

File: src/FragmentCache.php [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        $this->cache = $cache;
+        $this->webView = $webview;
+    }
+
+    /**
+     * Initializes the FragmentCache object.
+     *
+     * @throws InvalidArgumentException
+     */
+    public function init(): void
```
Comment (samdark) [inline src/FragmentCache.php]: Is there a better name?

File: src/FragmentCache.php [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    /**
+     * Generates a unique key used for storing the content in cache.
+     *
+     * The key generated depends on both {@see id} and {@see variations}.
+     *
+     * @return mixed a valid cache key
+     */
+    public function calculateKey()
```
Comment (samdark) [inline src/FragmentCache.php]: Is it public on purpose?

File: src/Spaceless.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Spaceless widget removes whitespace characters between HTML tags. Whitespaces within HTML tags or in a plain text are
+ * always left untouched.
+ *
+ * Usage example:
+ *
+ * ```php
+ * <body>
+ *     <?php Spaceless::begin(); ?>
```
Comment (samdark) [inline src/Spaceless.php]: Seems it's not current syntax.

File: src/ContentDecorator.php [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+use Yiisoft\View\Exception\ViewNotFoundException;
+use Yiisoft\View\WebView;
+use Yiisoft\Widget\Widget;
+
+/**
+ * ContentDecorator records all output between {@see begin()} and {@see end()]} calls, passes it to the given view file
+ * as `$content` and then echoes rendering result.
+ *
+ * ```php
+ * <?php ContentDecorator::begin()
```
Comment (samdark) [inline src/ContentDecorator.php]: Seems it's not current syntax.

File: src/Spaceless.php [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ * instead.
+ *
+ * @see http://tidy.sourceforge.net/
+ */
+class Spaceless extends Widget
+{
+    /**
+     * Starts capturing an output to be cleaned from whitespace characters between HTML tags.
+     */
+    public function init(): void
```
Comment (samdark) [inline src/Spaceless.php]: Is better name possible?

File: src/Spaceless.php [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+ * intention is to strip out extra whitespace characters between HTML tags in order to avoid browser rendering quirks
+ * in some circumstances (e.g. newlines between inline-block elements).
+ *
+ * Note, never use this method with `pre` or `textarea` tags. It's not that trivial to deal with such tags as it may
+ * seem at first sight. For this case you should consider using [HTML Tidy Project](http://tidy.sourceforge.net/)
+ * instead.
+ *
+ * @see http://tidy.sourceforge.net/
+ */
+class Spaceless extends Widget
```
Comment (samdark) [inline src/Spaceless.php]: Overall this one is pretty useless. I'd remove it.

File: src/Block.php (hunk 6 above)
Comment (terabytesoftw) [reply src/Block.php]: Done.

File: src/Block.php (hunk 7 above)
Comment (terabytesoftw) [reply src/Block.php]: start(), done.

File: src/Breadcrumbs.php (hunk 8 above)
Comment (terabytesoftw) [reply src/Breadcrumbs.php]: Done.

File: src/FragmentCache.php (hunk 10 above)
Comment (terabytesoftw) [reply src/FragmentCache.php]: done.

File: src/FragmentCache.php (hunk 11 above)
Comment (terabytesoftw) [reply src/FragmentCache.php]: yes, for tests check cache->has($key).

File: src/Spaceless.php (hunk 15 above)
Comment (terabytesoftw) [reply src/Spaceless.php]: remove spaceless done.

File: src/ContentDecorator.php (hunk 13 above)
Comment (terabytesoftw) [reply src/ContentDecorator.php]: update done.

File: src/Spaceless.php (hunk 12 above)
Comment (terabytesoftw) [reply src/Spaceless.php]: remove widget.

File: src/ContentDecorator.php (hunk 9 above)
Comment (terabytesoftw) [reply src/ContentDecorator.php]: done.

File: src/Breadcrumbs.php [hunk 16]
```diff
@@ -127,7 +129,7 @@ class Breadcrumbs extends Widget
      */
     public function run(): string
     {
-        if (empty($this->links)) {
+        if (!$this->links) {
```
Comment (xepozz) [inline src/Breadcrumbs.php]: Please don't compare any types as a boolean. If you want to check empty array or not, use `$var === []` or `empty($var)` or `count($var) === 0`. 
`!$var` too implicit expression IMHO.

Comment (terabytesoftw) [reply src/Breadcrumbs.php]: Done.

**Comment by samdark** [comment]
Merged! 👍 

Thank you.

**Pull request merged**